### PR TITLE
Update implementation to use new `Scenarios`

### PR DESCRIPTION
**Pull request by devonestes**
This commit updates this formatter to use the new `Scenario` structs in
Benchee core. There's probably some additional refactoring that could
be done, but it's such a small library as it is I didn't want to
over-do it with optimizations at this point.

I figured I'd start with an easy one before doing CSV and HTML :wink:

### CODE REVIEW

File: lib/benchee/formatters/json.ex [hunk 1]
```diff
@@ -49,57 +48,74 @@ defmodule Benchee.Formatters.JSON do
...
        end)
     |> Map.new
   end
 
-  @doc """
-  Transforms the benchmarking results of run times and staistics (for one input)
-  to JSON.
-
-  ## Examples
+  defp format_scenarios_for_input(scenarios) do
```
Comment (devonestes) [inline lib/benchee/formatters/json.ex]: I particularly like this function 😀

**Comment by PragTob** [comment]
I'm so sorry... I totally forgot about this one or may have looked at it once/marked as read. Ghost is very sorry:

img_20170723_143731

(I just noticed it as the other PR said "two down one to go)

File: lib/benchee/formatters/json.ex [hunk 2]
```diff
@@ -22,18 +22,17 @@ defmodule Benchee.Formatters.JSON do
...
   """
-  @spec output(Benchee.Suite.t) :: Benchee.Suite.t
-  def output(map)
-  def output(suite = %{configuration:
-                       %{formatter_options:
-                          %{json: %{file: filename}}}}) do
+  @spec output(Suite.t) :: Suite.t
+  def output(suite = %Suite{configuration: %Configuration{
+                        formatter_options: %{json: %{file: filename}}
+                      }}) do
```
Comment (PragTob) [inline lib/benchee/formatters/json.ex]: I still don't really know where to indent this,,, this seems a bit odd but no biggy :)

File: lib/benchee/formatters/json.ex [hunk 3]
```diff
@@ -49,57 +48,74 @@ defmodule Benchee.Formatters.JSON do
...
+      ...>          run_time_statistics: %Benchee.Statistics{
+      ...>            average:       500.0,
+      ...>            ips:           2000.0,
+      ...>            std_dev:       200.0,
+      ...>            std_dev_ratio: 0.4,
+      ...>            std_dev_ips:   800.0,
+      ...>            median:        450.0,
+      ...>            minimum:       200,
+      ...>            maximum:       900,
+      ...>            sample_size:   8
```
Comment (PragTob) [inline lib/benchee/formatters/json.ex]: :+1: formatting

File: lib/benchee/formatters/json.ex [hunk 4]
```diff
@@ -49,57 +48,74 @@ defmodule Benchee.Formatters.JSON do
...
-         {input, format_measurements(statistics_map, run_times_list)}
+
+  @spec format(Suite.t) :: %{Suite.key => String.t}
+  def format(%Suite{scenarios: scenarios}) do
+    scenarios
+    |> Enum.group_by(fn(scenario) -> scenario.input_name end)
+    |> Enum.map(fn({input, scenarios}) ->
+         {input, format_scenarios_for_input(scenarios)}
        end)
     |> Map.new
```
Comment (PragTob) [inline lib/benchee/formatters/json.ex]: also not too sure if we wanna keep the top level input_name key here. Imo we could just go all out and be like "we also change the structure here too" since we like our scenarios structure better, together with runtime_statistics et. al? I guess we would have to do a lot less work here then :)

Comment (PragTob) [inline lib/benchee/formatters/json.ex]: (speaking about gropuing, adding statistics etc.. just the sort order which we might need, but that's only relevant within one input... so it makes it more interesting. We might be able to live without it though? Not sure, question for the html formatter)

File: test/benchee/formatters/json_integration_test.exs [hunk 5]
```diff
@@ -2,58 +2,58 @@ defmodule Benchee.Formatters.JSONIntegrationTest do
...
-             formatters: [&Benchee.Formatters.JSON.output/1],
-             formatter_options: [json: [file: @file_path]]
-end
+  @file_path "test.json"
+  test "works just fine" do
+    basic_test time: 0.01,
+              warmup: 0.02,
+              formatters: [&Benchee.Formatters.JSON.output/1],
+              formatter_options: [json: [file: @file_path]]
+  end
```
Comment (PragTob) [inline test/benchee/formatters/json_integration_test.exs]: ugh that indentation.. damn sorry :(

File: test/benchee/formatters/json_test.exs [hunk 6]
```diff
@@ -4,42 +4,131 @@ defmodule Benchee.Formatters.JSONTest do
...
+      }
     }
-    filename = "test_input.json"
 
     try do
       capture_io fn ->
         return = Benchee.Formatters.JSON.output(suite)
         assert return == suite
       end
-      assert File.exists?(filename)
```
Comment (PragTob) [inline test/benchee/formatters/json_test.exs]: why don't we check for the filename anymore?

Comment (devonestes) [inline test/benchee/formatters/json_test.exs]: I thought that was covered in the integration tests [here](https://github.com/PragTob/benchee_json/blob/master/test/benchee/formatters/json_integration_test.exs#L28) so it seemed redundant to me. I'm happy to add it back in if you like!

File: lib/benchee/formatters/json.ex (hunk 2 above)
Comment (devonestes) [inline lib/benchee/formatters/json.ex]: I'm hoping that the new formatter will maybe solve all these problems for us! 😉

File: lib/benchee/formatters/json.ex (hunk 4 above)
Comment (devonestes) [inline lib/benchee/formatters/json.ex]: I'd be ok with it either way. Personally I haven't used this formatter yet (I pretty much only use the console formatter), so I'm probably not the best person to ask about what grouping would be most helpful for users. I sort of like the grouping for the console formatter by input (since it makes it easier for me to read), but I'm assuming that most people using the JSON formatter would be doing so not so a human could read it, but so that the data can be parsed by some other program. In that case, then maybe moving this output to match what we have internally (Scenarios), would make sense.

I can go either way with it. Maybe we can open an issue and try and gather some more information and have further discussion, and for now err on the side of non-breaking changes?

File: test/benchee/formatters/json_test.exs (hunk 6 above)
Comment (PragTob) [inline test/benchee/formatters/json_test.exs]: hm yeah, it might. It might not be too important.

File: lib/benchee/formatters/json.ex (hunk 4 above)
Comment (PragTob) [inline lib/benchee/formatters/json.ex]: Honestly, I don't think anyone uses the formatter much besides as a dependency of benchee_html. Somehow it even has less downloads than benchee_html. So, I think we can do mostly what we want. But yeah my expectation is that people using this would use it to read it into another tool (like R) or take the JSON and POST it to some web service, which is one of the end goals for the formatter (POSTing data to a web service/site that gathers results of benchmarks... maybe).

So, I think the new internal format would be better, but we can also put that off for another PR. Wdyt?

File: lib/benchee/formatters/json.ex (hunk 2 above)
Comment (PragTob) [inline lib/benchee/formatters/json.ex]: which new formatter... what did I miss?

Comment (devonestes) [inline lib/benchee/formatters/json.ex]: There's a Google Summer of Code project to write a code formatter for Elixir that José and Andrea are helping with! https://summerofcode.withgoogle.com/projects/#5205518009237504

File: lib/benchee/formatters/json.ex (hunk 4 above)
Comment (devonestes) [inline lib/benchee/formatters/json.ex]: Yeah, I agree that we can do that in a separate PR.

File: lib/benchee/formatters/json.ex (hunk 2 above)
Comment (PragTob) [inline lib/benchee/formatters/json.ex]: ah, nice!

**Comment by PragTob** [comment]
I think all discussion points are resolved :tada:
